Lykke.B2c2Client/B2С2WebSocketClient.cs

```diff
@@ -0,0 +1,633 @@
+                lock (_lockTimestamp)
+                {
+                    _lastSuccessPriceMessageTimestamp = value;
+                }
+            }
+        }
+
+        public B2С2WebSocketClient(B2C2ClientSettings settings, ILogFactory logFactory)
+        {
+            if (settings == null) throw new NullReferenceException(nameof(settings));
+            var url = settings.Url;
+            var authorizationToken = settings.AuthorizationToken;
+            if (string.IsNullOrWhiteSpace(url) || !Uri.TryCreate(url, UriKind.Absolute, out _))
+                throw new ArgumentOutOfRangeException(nameof(url));
+            if (string.IsNullOrWhiteSpace(authorizationToken)) throw new ArgumentOutOfRangeException(nameof(authorizationToken));
+            if (logFactory == null) throw new NullReferenceException(nameof(logFactory));
+
+            _baseUri = url[url.Length-1] == '/' ? url.Substring(0, url.Length - 1) : url;
+            _authorizationToken = authorizationToken;
+            _log = logFactory.CreateLog(this);
+            _clientWebSocket = new ClientWebSocket();
+            _awaitingSubscriptions = new ConcurrentDictionary<string, Subscription>();
+            _instrumentsHandlers = new ConcurrentDictionary<string, Func<PriceMessage, Task>>();
+            _instrumentsLevels = new ConcurrentDictionary<string, decimal[]>();
+            _awaitingUnsubscriptions = new ConcurrentDictionary<string, Subscription>();
+            _tradableInstruments = new List<string>();
+            _cancellationTokenSource = new CancellationTokenSource();
+            _reconnectIfNeededTrigger = new TimerTrigger(nameof(B2С2WebSocketClient), new TimeSpan(0, 0, 0, 30), logFactory, ReconnectIfNeeded);
+            _reconnectIfNeededTrigger.Start();
+        }
+
+        public Task SubscribeAsync(string instrument, decimal[] levels, Func<PriceMessage, Task> handler,
+            CancellationToken ct = default(CancellationToken))
+        {
+            ThrowIfSubscriptionIsAlreadyExist(instrument);
+
+            ConnectIfNeeded(ct);
+
+            var tag = Guid.NewGuid().ToString();
+
+            _log.Info($"Attempt to subscribe to order book updates, instrument: '{instrument}'.", tag);
+
+            var subscribeRequest = new SubscribeRequest { Instrument = instrument, Levels = levels, Tag = tag };
+            SendMessageToWebSocket(subscribeRequest, ct).GetAwaiter().GetResult();
+
+            // Save subscription state
+            var taskCompletionSource = new TaskCompletionSource<int>();
+            lock (_sync)
+            {
+                _awaitingSubscriptions[instrument] = new Subscription(tag, taskCompletionSource, handler);
+                _instrumentsLevels[instrument] = levels;
+            }
+
+            // Throw exception if time out
+            #pragma warning disable CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+            Task.Run(async () =>
```

Comment:
here you can use more simple code:
var successTask = await Task.WhenAny(taskCompletionSource.Task, Task.Delay(_timeOut, ct)

if (successTask != taskCompletionSource.Task)
{
...
throw new B2c2WebSocketException($"Subscription timeout for {instrument}.")
}
...
return;